File: agent/internal/tool/http_request.go
package tool

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func httpRequestTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "http_request",
			Description: "Make a custom HTTP request to any URL. Use this for testing APIs, fetching data, or checking endpoints.",
			Category:    "network",
			Parameters: map[string]api.ParamDef{
				"url":     {Type: "string", Description: "Full URL to request", Required: true},
				"method":  {Type: "string", Description: "HTTP method", Required: false, Default: "GET", Enum: []string{"GET", "POST", "PUT", "DELETE", "PATCH", "HEAD"}},
				"headers": {Type: "string", Description: "JSON object of custom headers", Required: false},
				"body":    {Type: "string", Description: "Request body (for POST/PUT/PATCH)", Required: false},
				"timeout": {Type: "number", Description: "Timeout in seconds", Required: false, Default: 15},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			url, _ := params["url"].(string)
			if url == "" {
				return nil, fmt.Errorf("url is required")
			}
			method, _ := params["method"].(string)
			if method == "" {
				method = "GET"
			}
			timeoutSec, _ := params["timeout"].(float64)
			if timeoutSec == 0 {
				timeoutSec = 15
			}

			var bodyReader io.Reader
			if bodyStr, ok := params["body"].(string); ok && bodyStr != "" {
				bodyReader = strings.NewReader(bodyStr)
			}

			ctx, cancel := context.WithTimeout(ctx, time.Duration(timeoutSec)*time.Second)
			defer cancel()

			req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
			if err != nil {
				return nil, fmt.Errorf("failed to create request: %w", err)
			}

			if headersStr, ok := params["headers"].(string); ok && headersStr != "" {
				var headers map[string]string
				if err := json.Unmarshal([]byte(headersStr), &headers); err == nil {
					for k, v := range headers {
						req.Header.Set(k, v)
					}
				}
			}

			start := time.Now()
			resp, err := http.DefaultClient.Do(req)
			if err != nil {
				return nil, fmt.Errorf("request failed: %w", err)
			}
			defer resp.Body.Close()

			respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 1024*512))

			return map[string]any{
				"url":          url,
				"status_code":  resp.StatusCode,
				"status":       resp.Status,
				"content_type": resp.Header.Get("Content-Type"),
				"body":         string(respBody),
				"body_size":    len(respBody),
				"duration_ms":  time.Since(start).Milliseconds(),
			}, nil
		},
	}
}

File: src/services/tools/network/httpRequest.ts
import { z } from 'zod';
import type { ToolDef } from '../types';

export const httpRequestTool: ToolDef = {
  name: 'http_request',
  description: 'Make an HTTP request to an external API or URL. Supports GET, POST, PUT, PATCH, DELETE methods.',
  category: 'network',
  inputSchema: z.object({
    url: z.string().url().describe('The full URL to send the request to.'),
    method: z.enum(['GET', 'POST', 'PUT', 'PATCH', 'DELETE']).optional().default('GET').describe('HTTP method.'),
    headers: z.record(z.string()).optional().describe('Optional HTTP headers as key-value pairs.'),
    body: z.string().optional().describe('Request body as a string (JSON, text, etc.). Automatically sets Content-Type for JSON.'),
    timeout: z.number().optional().default(15000).describe('Timeout in milliseconds.'),
  }),
  execute: async ({ url, method, headers, body, timeout }) => {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout);

    try {
      const reqHeaders: Record<string, string> = { ...headers };
      if (body && !reqHeaders['Content-Type']) {
        try {
          JSON.parse(body);
          reqHeaders['Content-Type'] = 'application/json';
        } catch {
          reqHeaders['Content-Type'] = 'text/plain';
        }
      }

      const res = await fetch(url, {
        method,
        headers: reqHeaders,
        body: body || undefined,
        signal: controller.signal,
      });

      const contentType = res.headers.get('content-type') || '';
      let data: any;
      if (contentType.includes('application/json')) {
        data = await res.json();
      } else {
        data = await res.text();
      }

      return {
        status: res.status,
        statusText: res.statusText,
        headers: Object.fromEntries(res.headers.entries()),
        data,
      };
    } catch (error: any) {
      if (error.name === 'AbortError') {
        throw new Error(`Request timed out after ${timeout}ms`);
      }
      throw new Error(`HTTP request failed: ${error.message}`);
    } finally {
      clearTimeout(timer);
    }
  },
};
